src/diffusers/models/controlnetxs.py

```diff
@@ -0,0 +1,1009 @@
+            h_base = h_base + next(it_down_convs_out)(h_ctrl) * next(scales)  # D - add ctrl -> base
+            hs_base.append(h_base)
+            hs_ctrl.append(h_ctrl)
+
+        # 2 - mid
+        h_ctrl = torch.cat([h_ctrl, next(it_down_convs_in)(h_base)], dim=1)  # A - concat base -> ctrl
+        for m_base, m_ctrl in zip(base_mid_subblocks, ctrl_mid_subblocks):
+            h_base = m_base(h_base, temb, cemb, attention_mask, cross_attention_kwargs)  # B - apply base subblock
+            h_ctrl = m_ctrl(h_ctrl, temb, cemb, attention_mask, cross_attention_kwargs)  # C - apply ctrl subblock
+        h_base = h_base + self.middle_block_out(h_ctrl) * next(scales)  # D - add ctrl -> base
+
+        # 3 - up
+        for i, m_base in enumerate(base_up_subblocks):
+            h_base = h_base + next(it_up_convs_out)(hs_ctrl.pop()) * next(scales)  # add info from ctrl encoder
+            h_base = torch.cat([h_base, hs_base.pop()], dim=1)  # concat info from base encoder+ctrl encoder
+            h_base = m_base(h_base, temb, cemb, attention_mask, cross_attention_kwargs)
+
+        h_base = base_model.conv_norm_out(h_base)
+        h_base = base_model.conv_act(h_base)
+        h_base = base_model.conv_out(h_base)
+
+        if not return_dict:
+            return h_base
+
+        return ControlNetXSOutput(sample=h_base)
+
+    def make_zero_conv(self, in_channels, out_channels=None):
+        # keep running track of channels sizes
+        self.in_channels = in_channels
+        self.out_channels = out_channels or in_channels
+
+        return zero_module(nn.Conv2d(in_channels, out_channels, 1, padding=0))
+
+    @torch.no_grad()
+    def _check_if_vae_compatible(self, vae: AutoencoderKL):
+        condition_downscale_factor = 2 ** (len(self.config.conditioning_block_sizes) - 1)
+
+        # Multiply by 2, as otherwise we have channel with sizes = 1 after vae encoding, which confuses PyTorch.
+        # Alternativy, we could set the vae to eval mode.
+        in_size = condition_downscale_factor * 2
+
+        rand_tensor = torch.rand((1, 3, in_size, in_size)).to(vae.device, dtype=vae.dtype)
+
+        encoded_tensor = vae.encode(rand_tensor)
+        if hasattr(encoded_tensor, "latent_dist"):
+            encoded_tensor = encoded_tensor.latent_dist.sample()
+        elif hasattr(encoded_tensor, "latents"):
+            encoded_tensor = encoded_tensor.latents
+        else:
+            raise ValueError(f"Output of {type(vae)} has neither `latents` nor `latent_dist` as attribute.")
+
+        out_size = encoded_tensor.shape[-1]
+
+        vae_downscale_factor = in_size / out_size
+
+        compatible = condition_downscale_factor == vae_downscale_factor
+
+        return compatible, condition_downscale_factor, vae_downscale_factor
+
+
+class EmbedSequential(nn.ModuleList):
+    """Sequential module passing embeddings (time and conditioning) to children if they support it."""
```

Comment:
Could we try to get rid of this class as it's not really in line with the other controlnet implementation. It would be great to try to explicitly call the layers that are needed at a certain position

Reply:
Iiuc, we need `EmbedSequential` because it represents the concept of a "subblock". (To make this clearer, I've renamed the class `SubBlock`.) A subblock is the largest piece of the control or base unet, that can be executed independently of the other unet respectively.

Unlike ControlNet, in ControlNet-XS information is exchanged between base and control model on a subblock level. A `CrossAttnDownBlock2D`, for example, is split into 3 subblocks: ResNet+Attention, ResNet+Attention, and Downsampler. After each subblock, information is exchanged from control to base and vice versa.

In the original ControlNet, because the entire control model is executed independently of the base model, it's not necessary to have the concept of a subblock.

What do you think?